.github/workflows/scripts/ghstack-perm-check.py

```diff
@@ -152,19 +159,91 @@ def rebase(self):
                 f"User {actor} must be a maintainer {self.REPO} to rebase someone else's PR",
             )
 
+
+class GHStack(ChatOps):
+    def __init__(self, session, event):
+        super().__init__(session, event)
+        self.pr_numbers = self.ghstack_pr_numbers()
+
+    def get_pr_ref(self):
+        return self.PR["head"]["ref"].replace("/head", "/orig")
+
+    def ghstack_pr_numbers(self):
+        proc = subprocess.Popen(
+            "git log FETCH_HEAD...$(git merge-base FETCH_HEAD origin/main)",
+            stdout=subprocess.PIPE,
+            shell=True,
+        )
+        out, _ = proc.communicate()
+        self.must(proc.wait() == 0, "`git log` command failed!")
+
+        pr_numbers = re.findall(
+            r"Pull Request resolved: https://github.com/.*?/pull/([0-9]+)",
+            out.decode("utf-8"),
+        )
+        pr_numbers = list(map(int, pr_numbers))
+        self.must(
+            pr_numbers and pr_numbers[0] == self.NUMBER,
+            "Extracted PR numbers not seems right!",
+        )
+        return pr_numbers
+
+    def land(self):
+        # Enforce requirements for every PR in the stack
+        for n in self.pr_numbers:
+            # Checking Approvals
+            self.land_prerequisites(n)
+
+        print(":: All PRs are ready to be landed!")
+
+    def rebase(self):
+        self.rebase_prerequisites()
         print(":: All PRs are ready to be rebased!")
 
 
+class GitHub(ChatOps):
+    def get_pr_ref(self):
+        return f"pull/{self.NUMBER}/head"
+
+    def must(self, cond, msg):
+        must(self.gh, self.REPO, self.NUMBER, cond, msg)
+
+    def land(self):
+        # Enforce requirements for the PR
+        self.land_prerequisites()
+        print(":: The PR is ready to be landed!")
+
+    def rebase(self):
+        self.rebase_prerequisites()
+        print(":: The PR is ready to be rebased!")
+
+
 def main(check):
     # Setup the wrapper class
     EV = json.loads(sys.stdin.read())
-    ghstack_checks = GHStackChecks(os.environ["GITHUB_TOKEN"], EV)
+    session = get_session(os.environ["GITHUB_TOKEN"])
+    chatops = None
+    isgh = is_ghstack(session, EV)
+
+    if isgh:
+        chatops = GHStack(session, EV)
+    else:
+        chatops = GitHub(session, EV)
+
+    # Set a number of useful environment variables
+    env_file = os.getenv("GITHUB_ENV")
```

Comment:
this code is a bit confusing... it says it's setting environment variables but I believe it's writing key=value pairs to a file? the filename is in GITHUB_ENV but where is GITHUB_ENV set? I also don't see how these values are actually read back into the environment, this script is just run in python?

Did you test that this behavior actually picks up the relevant variables correctly?

Reply:
This code relies on the GitHub action functionality described in [this doc](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable).

GITHUB_ENV contains the path to a file that is loaded before each step in a job. 
I can improve the comment in the code to make it more clear what happens.

I did test this code in my sandbox repo.

Reply:
ah I see. And so there's some git command run before `.github/workflows/scripts/ghstack-perm-check.py` that sets the GITHUB_ENV to the relevant file name? what command does that?

I think a reference to the doc might be helpful for future readers but this makes more sense to me now.